### PR TITLE
Lsm merge aggressive

### PR DESCRIPTION
**Pull request by agorrod**
refs WT-1955

### CODE REVIEW

**Comment by agorrod** [comment]
@sueloverso Could you review this change please? I'm seeing much more stable runs of the wtperf medium multi LSM workload with this code, which leads to better throughput.

For example, on the branch I'm seeing:

```
2164967 reads, 0 inserts, 2164965 updates, 0 checkpoints in 5 secs (30 total secs)
2103833 reads, 0 inserts, 2103405 updates, 0 checkpoints in 5 secs (35 total secs)
1883905 reads, 0 inserts, 1884622 updates, 0 checkpoints in 5 secs (40 total secs)
1973095 reads, 0 inserts, 1959985 updates, 0 checkpoints in 5 secs (45 total secs)
1988264 reads, 0 inserts, 1786102 updates, 0 checkpoints in 5 secs (50 total secs)
1643348 reads, 0 inserts, 1576455 updates, 0 checkpoints in 5 secs (55 total secs)
2017243 reads, 0 inserts, 1700232 updates, 0 checkpoints in 5 secs (60 total secs)
1695125 reads, 0 inserts, 1676507 updates, 0 checkpoints in 5 secs (65 total secs)
1771364 reads, 0 inserts, 1352547 updates, 0 checkpoints in 5 secs (70 total secs)
1791629 reads, 0 inserts, 1532012 updates, 0 checkpoints in 5 secs (75 total secs)
1847110 reads, 0 inserts, 1587092 updates, 0 checkpoints in 5 secs (80 total secs)
2278752 reads, 0 inserts, 1664166 updates, 0 checkpoints in 5 secs (85 total secs)
2005394 reads, 0 inserts, 1405126 updates, 0 checkpoints in 5 secs (90 total secs)
1869856 reads, 0 inserts, 1310296 updates, 0 checkpoints in 5 secs (95 total secs)
1858649 reads, 0 inserts, 1458463 updates, 0 checkpoints in 5 secs (100 total secs)
1721571 reads, 0 inserts, 1217470 updates, 0 checkpoints in 5 secs (105 total secs)
1741439 reads, 0 inserts, 1269533 updates, 0 checkpoints in 5 secs (110 total secs)
1512408 reads, 0 inserts, 1025750 updates, 0 checkpoints in 5 secs (115 total secs)
1648530 reads, 0 inserts, 1372432 updates, 0 checkpoints in 5 secs (120 total secs)
1687500 reads, 0 inserts, 1165594 updates, 0 checkpoints in 5 secs (125 total secs)
1720646 reads, 0 inserts, 1168197 updates, 0 checkpoints in 5 secs (130 total secs)
1667410 reads, 0 inserts, 1138620 updates, 0 checkpoints in 5 secs (135 total secs)
1671801 reads, 0 inserts, 1253277 updates, 0 checkpoints in 5 secs (140 total secs)
1874970 reads, 0 inserts, 1268652 updates, 0 checkpoints in 5 secs (145 total secs)
1703028 reads, 0 inserts, 1361564 updates, 0 checkpoints in 5 secs (150 total secs)
1793996 reads, 0 inserts, 1207449 updates, 0 checkpoints in 5 secs (155 total secs)
1773936 reads, 0 inserts, 1219111 updates, 0 checkpoints in 5 secs (160 total secs)
1717693 reads, 0 inserts, 1146072 updates, 0 checkpoints in 5 secs (165 total secs)
1742015 reads, 0 inserts, 1184328 updates, 0 checkpoints in 5 secs (170 total secs)
1622205 reads, 0 inserts, 1118374 updates, 0 checkpoints in 5 secs (175 total secs)
1559051 reads, 0 inserts, 1132897 updates, 0 checkpoints in 5 secs (180 total secs)
1591576 reads, 0 inserts, 1066388 updates, 0 checkpoints in 5 secs (185 total secs)
1291907 reads, 0 inserts, 1021933 updates, 0 checkpoints in 5 secs (190 total secs)
1575362 reads, 0 inserts, 1054079 updates, 0 checkpoints in 5 secs (195 total secs)
1263089 reads, 0 inserts, 854613 updates, 0 checkpoints in 5 secs (200 total secs)
```

On develop I see:

```
2076568 reads, 0 inserts, 2076511 updates, 0 checkpoints in 5 secs (30 total secs)
1194195 reads, 0 inserts, 1194347 updates, 0 checkpoints in 5 secs (35 total secs)
1684896 reads, 0 inserts, 1684776 updates, 0 checkpoints in 5 secs (40 total secs)
1564545 reads, 0 inserts, 1564572 updates, 0 checkpoints in 5 secs (45 total secs)
514387 reads, 0 inserts, 675773 updates, 0 checkpoints in 5 secs (50 total secs)
0 reads, 0 inserts, 390932 updates, 0 checkpoints in 5 secs (55 total secs)
1268178 reads, 0 inserts, 327260 updates, 0 checkpoints in 5 secs (60 total secs)
1690308 reads, 0 inserts, 1787328 updates, 0 checkpoints in 5 secs (65 total secs)
1316210 reads, 0 inserts, 1599480 updates, 0 checkpoints in 5 secs (70 total secs)
1581332 reads, 0 inserts, 1589606 updates, 0 checkpoints in 5 secs (75 total secs)
1346652 reads, 0 inserts, 1288904 updates, 0 checkpoints in 5 secs (80 total secs)
2154416 reads, 0 inserts, 177583 updates, 0 checkpoints in 5 secs (85 total secs)
1366913 reads, 0 inserts, 1658 updates, 0 checkpoints in 5 secs (90 total secs)
913952 reads, 0 inserts, 630956 updates, 0 checkpoints in 5 secs (95 total secs)
1453948 reads, 0 inserts, 1247293 updates, 0 checkpoints in 5 secs (100 total secs)
2462396 reads, 0 inserts, 2119294 updates, 0 checkpoints in 5 secs (105 total secs)
1613149 reads, 0 inserts, 1930394 updates, 0 checkpoints in 5 secs (110 total secs)
1909337 reads, 0 inserts, 1608941 updates, 0 checkpoints in 5 secs (115 total secs)
1472948 reads, 0 inserts, 1240812 updates, 0 checkpoints in 5 secs (120 total secs)
1532822 reads, 0 inserts, 440574 updates, 0 checkpoints in 5 secs (125 total secs)
0 reads, 0 inserts, 0 updates, 0 checkpoints in 5 secs (130 total secs)
627036 reads, 0 inserts, 473078 updates, 0 checkpoints in 5 secs (135 total secs)
1141972 reads, 0 inserts, 1103475 updates, 0 checkpoints in 5 secs (140 total secs)
1755218 reads, 0 inserts, 1790043 updates, 0 checkpoints in 5 secs (145 total secs)
2484436 reads, 0 inserts, 1755146 updates, 0 checkpoints in 5 secs (150 total secs)
2391767 reads, 0 inserts, 2015459 updates, 0 checkpoints in 5 secs (155 total secs)
2481024 reads, 0 inserts, 2134618 updates, 0 checkpoints in 5 secs (160 total secs)
2542760 reads, 0 inserts, 751864 updates, 0 checkpoints in 5 secs (165 total secs)
2983680 reads, 0 inserts, 0 updates, 0 checkpoints in 5 secs (170 total secs)
1077712 reads, 0 inserts, 1010368 updates, 0 checkpoints in 5 secs (175 total secs)
2397327 reads, 0 inserts, 1794227 updates, 0 checkpoints in 5 secs (180 total secs)
2633406 reads, 0 inserts, 1971795 updates, 0 checkpoints in 5 secs (185 total secs)
922490 reads, 0 inserts, 695419 updates, 0 checkpoints in 5 secs (190 total secs)
1321587 reads, 0 inserts, 615784 updates, 0 checkpoints in 5 secs (195 total secs)
3287854 reads, 0 inserts, 38618 updates, 0 checkpoints in 5 secs (200 total secs)
```

**Comment by agorrod** [comment]
I forgot to add a description of the problem. The issue is that the current code in LSM that selects which chunks to include a merge isn't looking closely enough at the generation of the chunks it's choosing. Each time a merge completes the resultant chunk is assigned a generation of previous highest generation plus one.

With a fairly normal fan-out the difference in size between different generations should be an order of magnitude. So merging chunks across generations should be exceptional (only when we can't find other merges to do).

The code was intended to work that way, but there wasn't enough logic to enforce it.

**Comment by sueloverso** [comment]
@agorrod I don't think this is ready to merge - see my comment and results in WT-1955.

File: src/include/lsm.h [hunk 1]
```diff
@@ -219,11 +225,12 @@ struct __wt_lsm_tree {
 	uint32_t merge_aggressiveness;	/* Increase amount of work per merge */
 
 #define	WT_LSM_TREE_ACTIVE		0x01	/* Workers are active */
-#define	WT_LSM_TREE_COMPACTING		0x02	/* Tree being compacted */
-#define	WT_LSM_TREE_MERGES		0x04	/* Tree should run merges */
-#define	WT_LSM_TREE_NEED_SWITCH		0x08	/* New chunk needs creating */
-#define	WT_LSM_TREE_OPEN		0x10	/* The tree is open */
-#define	WT_LSM_TREE_THROTTLE		0x20	/* Throttle updates */
+#define	WT_LSM_TREE_AGGRESSIVE_TIMER	0x02	/* Workers are active */
```
Comment (sueloverso) [inline src/include/lsm.h]: Please update the comment, which appears copied from `WT_LSM_TREE_ACTIVE`.

File: src/lsm/lsm_manager.c [hunk 2]
```diff
@@ -458,7 +418,7 @@ __lsm_manager_run_server(WT_SESSION_IMPL *session)
 			    lsm_tree->nchunks > 1) ||
 			    (lsm_tree->queue_ref == 0 &&
 			    lsm_tree->nchunks > 1) ||
-			    (lsm_tree->merge_aggressiveness > 3 &&
+			    (lsm_tree->merge_aggressiveness > 2 &&
```
Comment (sueloverso) [inline src/lsm/lsm_manager.c]: Should this `2` be `WT_LSM_AGGRESSIVE_THRESHOLD`?  Are these numbers related?  If not, maybe another macro value that explains what this number is?

File: src/lsm/lsm_manager.c [hunk 3]
```diff
@@ -469,7 +429,8 @@ __lsm_manager_run_server(WT_SESSION_IMPL *session)
 				    session, WT_LSM_WORK_FLUSH, 0, lsm_tree));
 				WT_ERR(__wt_lsm_manager_push_entry(
 				    session, WT_LSM_WORK_BLOOM, 0, lsm_tree));
-				WT_ERR(__wt_verbose(session, WT_VERB_LSM,
+				WT_ERR(__wt_verbose(session,
+				    WT_VERB_LSM_MANAGER,
```
Comment (sueloverso) [inline src/lsm/lsm_manager.c]: I think having a separate category is a good idea.  There are probably more verbose messages that would be useful in this category.  Reconfig to start/stop threads and clearing all work units from a particular tree are two areas that come to mind.

File: src/lsm/lsm_merge.c [hunk 4]
```diff
@@ -40,6 +40,80 @@ __wt_lsm_merge_update_tree(WT_SESSION_IMPL *session,
...
+	 */
+	if (lsm_tree->chunks_flushed <= lsm_tree->merge_min)
+		return (__lsm_merge_aggressive_clear(lsm_tree));
+
+	/* Start the timer if it isn't running */
+	if (!F_ISSET(lsm_tree, WT_LSM_TREE_AGGRESSIVE_TIMER))
+		return (__wt_epoch(session, &lsm_tree->merge_aggressive_ts));
+
+	WT_RET(__wt_epoch(session, &now));
+	time_since_last_merge =
```
Comment (sueloverso) [inline src/lsm/lsm_merge.c]: Can we please call this `msec_since_last_merge`?

File: src/lsm/lsm_merge.c [hunk 5]
```diff
@@ -40,6 +40,80 @@ __wt_lsm_merge_update_tree(WT_SESSION_IMPL *session,
...
+
+	/*
+	 * Only get aggressive if a reasonable number of flushes have been
+	 * completed since opening the tree.
+	 */
+	if (lsm_tree->chunks_flushed <= lsm_tree->merge_min)
+		return (__lsm_merge_aggressive_clear(lsm_tree));
+
+	/* Start the timer if it isn't running */
+	if (!F_ISSET(lsm_tree, WT_LSM_TREE_AGGRESSIVE_TIMER))
```
Comment (sueloverso) [inline src/lsm/lsm_merge.c]: There is no place that sets this flag.  Where is it supposed to get set?

File: src/lsm/lsm_merge.c [hunk 6]
```diff
@@ -40,6 +40,80 @@ __wt_lsm_merge_update_tree(WT_SESSION_IMPL *session,
...
+		return (__wt_epoch(session, &lsm_tree->merge_aggressive_ts));
+
+	WT_RET(__wt_epoch(session, &now));
+	time_since_last_merge =
+	    WT_TIMEDIFF(now, lsm_tree->merge_aggressive_ts) / WT_MILLION;
+
+	/*
+	 * Never get aggressive before we should have created enough chunks
+	 * for a new merge.
```
Comment (sueloverso) [inline src/lsm/lsm_merge.c]: This comment is awkwardly worded.  Maybe remove `should`?

File: src/lsm/lsm_merge.c [hunk 7]
```diff
@@ -203,22 +287,32 @@ __lsm_merge_span(WT_SESSION_IMPL *session, WT_LSM_TREE *lsm_tree,
 
 	WT_ASSERT(session,
 	    nchunks == 0 || (chunk != NULL && youngest != NULL));
+
```
Comment (sueloverso) [inline src/lsm/lsm_merge.c]: Did you intend to add this blank line?

File: src/lsm/lsm_merge.c [hunk 8]
```diff
@@ -299,8 +393,12 @@ __wt_lsm_merge(WT_SESSION_IMPL *session, WT_LSM_TREE *lsm_tree, u_int id)
...
 		for (verb = start_chunk; verb <= end_chunk; verb++)
 			WT_ERR(__wt_verbose(session, WT_VERB_LSM,
-			    "%s: Chunk[%u] id %u",
-			    lsm_tree->name, verb, lsm_tree->chunk[verb]->id));
+			    "Merging %s: Chunk[%u] id %u, gen: %" PRIu32
+			    ", size: %" PRIu64 ", records: %" PRIu64,
+			    lsm_tree->name, verb, lsm_tree->chunk[verb]->id,
+			    lsm_tree->chunk[verb]->generation,
+			    lsm_tree->chunk[verb]->size,
+			    lsm_tree->chunk[verb]->count));
```
Comment (sueloverso) [inline src/lsm/lsm_merge.c]: This is a great update to the message.  Much more useful information.

**Comment by sueloverso** [comment]
@agorrod I am mostly good with this change.  Most of my comments are cosmetic and I would have just made the changes myself and merged, except there is one missing line to set the flag and I'm not sure where that should be.  I'll comment on my results in the ticket.

**Comment by agorrod** [comment]
@sueloverso I've incorporated your review comments (thanks). I'm going to merge now, because I want to see the results in the performance tests. It would be good if you can take a final look through the last commits to make sure I addressed the issues you had.
